jdbc/src/main/java/org/jclouds/jdbc/service/JdbcService.java

```diff
@@ -16,54 +16,242 @@
+         BlobToBlobEntity blobToBlobEntity) {
       this.containerRepository = containerRepository;
+      this.blobRepository = blobRepository;
+      this.chunkRepository = chunkRepository;
+      this.blobToBlobEntity = blobToBlobEntity;
    }
 
    @Transactional
-   public void createContainer(String name, ContainerAccess access) {
-      containerRepository.create(Container.builder().name(name).containerAccess(access).build());
+   public void createContainer(String containerName, ContainerAccess access) {
+      containerRepository.create(ContainerEntity.builder().name(containerName).containerAccess(access).build());
    }
 
    @Transactional
-   public void createContainer(String name) {
-      containerRepository.create(Container.builder().name(name).build());
+   public void createContainer(String containerName) {
+      createContainer(containerName, null);
    }
 
    @Transactional
-   public List<Container> findAllContainers() {
+   public List<ContainerEntity> findAllContainers() {
       return containerRepository.findAllContainers();
    }
 
    @Transactional
-   public Container findContainerByName(String name) {
-      return containerRepository.findContainerByName(name);
+   public ContainerEntity findContainerByName(String containerName) {
+      return containerRepository.findContainerByName(containerName);
    }
 
    @Transactional
-   public void deleteContainerByName(String name) {
-      containerRepository.deleteContainerByName(name);
+   public void deleteContainerByName(String containerName) {
+      containerRepository.deleteContainerByName(containerName);
    }
 
    @Transactional
-   public void setContainerAccessByName(String name, ContainerAccess access) {
-      Container container = containerRepository.findContainerByName(name);
-      container.setContainerAccess(access);
-      containerRepository.save(container);
+   public void setContainerAccessByName(String containerName, ContainerAccess access) {
+      ContainerEntity containerEntity = containerRepository.findContainerByName(containerName);
+      containerEntity.setContainerAccess(access);
+      containerRepository.save(containerEntity);
    }
+
+   @Transactional
+   public boolean blobExists(String containerName, String key) {
+      return findBlobById(containerName, key) != null;
+   }
+
+   @Transactional(rollbackOn = IOException.class)
+   public BlobEntity createOrModifyBlob(String containerName, Blob blob, BlobAccess blobAccess) throws IOException {
+      String key = blob.getMetadata().getName();
+      Date creationDate = null;
+      BlobEntity oldBlobEntity = findBlobById(containerName, key);
+      if (oldBlobEntity != null) {
+         creationDate = oldBlobEntity.getCreationDate();
+         deleteBlob(containerName, key);
+      }
+      BlobEntity blobEntity = blobToBlobEntity.apply(blob);
+      blobEntity.getPayload().setChunks(storeData(blob.getPayload().openStream()));
```

Comment:
The InputStream is closed in the storeData method :)

I think we need to check the integrity after the data is copied to the database. I think we need to either set things up in the database to make it return the hash of the copied data, or re-read the data from it and calculate it. Any other ideas?